### PR TITLE
Add config option for disabling the events daemon

### PR DESCRIPTION
**Pull request by zachmullen**
@mgrauer PTAL. This allows a config file switch to be set such that `girder.events.daemon` simply becomes an alias for the normal, synchronous `girder.events`, which makes us no longer depend on having a background thread on the server.

### CODE REVIEW

File: girder/utility/config.py [hunk 1]
```diff
@@ -40,15 +40,15 @@ def _loadConfigsByPrecedent():
...
-        os.path.join(PACKAGE_DIR, 'conf', 'girder.dist.cfg'))
-    configPaths.append(
-        os.path.join(PACKAGE_DIR, 'conf', 'girder.local.cfg'))
-    configPaths.append(
-        os.path.join('/etc', 'girder.cfg'))
-    configPaths.append(
-        os.path.join(os.path.expanduser('~'), '.girder', 'girder.cfg'))
+    configPaths = [os.path.join(PACKAGE_DIR, 'conf', 'girder.dist.cfg')]
+
+    if 'GIRDER_TEST_DB' not in os.environ:
```
Comment (mgrauer) [inline girder/utility/config.py]: why this change?  is it necessary for the new test?

File: girder/utility/server.py [hunk 2]
```diff
@@ -103,8 +103,9 @@ def configureServer(test=False, plugins=None, curConfig=None):
     root = webroot.Webroot()
     api_main.addApiToNode(root)
 
-    cherrypy.engine.subscribe('start', girder.events.daemon.start)
-    cherrypy.engine.subscribe('stop', girder.events.daemon.stop)
+    if not curConfig['server'].get('disable_event_daemon', False):
+        cherrypy.engine.subscribe('start', girder.events.daemon.start)
```
Comment (mgrauer) [inline girder/utility/server.py]: add a comment explaining why you don't do this, in this case

File: tests/cases/events_test.py [hunk 3]
```diff
@@ -136,3 +139,14 @@ def callback(event):
...
+    def testDisableDaemon(self):
+        self.assertTrue(hasattr(events.daemon, 'start'))
+
+        # Disable the event daemon and reload the module
+        config.getConfig()['server']['disable_event_daemon'] = True
+        del sys.modules['girder.events']
+        eventsModule = importlib.import_module('girder.events')
+
+        self.assertFalse(hasattr(eventsModule.daemon, 'start'))
+        self.assertTrue(callable(eventsModule.daemon.trigger))
```
Comment (mgrauer) [inline tests/cases/events_test.py]: is it possible to run `testAsyncEvents` with `disable_event_daemon` set?  it would be nice to test the behavior in some way in the case of no async thread, as clients can still call daemon methods.

File: girder/events.py [hunk 4]
```diff
@@ -266,4 +268,9 @@ def trigger(eventName, info=None, pre=None, async=False):
 
 _deprecated = {}
 _mapping = {}
-daemon = AsyncEventsThread()
+
+if config.getConfig()['server'].get('disable_event_daemon', False):
+    # If daemon is disabled, just use the current module to run those event synchronously
```
Comment (mgrauer) [inline girder/events.py]: events

**Comment by zachmullen** [comment]
> it might be better to raise an exception in the case where an async event is attempted when running without the async thread

In principal I agree, but there are core actions that call `girder.events.daemon.trigger`, and the `bind` part of things does not distinguish between binding synchronously or asynchronously, so I don't think that's possible. For instance, uploading a file would always raise an exception in that case. A compromise solution might be to print a warning message if someone has bound to one of these events that is normally triggered on the daemon, but is being triggered synchronously.

**Comment by zachmullen** [comment]
I worded that badly: it's possible, and we could throw an exception in that case, but I think a warning would be a little better than just shutting down the request.

File: girder/utility/config.py (hunk 1 above)
Comment (zachmullen) [inline girder/utility/config.py]: I noticed that the test execution was being affected by how I had my girder.local.cfg set up, which is not desirable, so I went ahead and fixed that issue here.

**Comment by zachmullen** [comment]
@mgrauer I changed this around a bit. Added warning in the case you mentioned, and improved the testing.

File: tests/cases/events_test.py [hunk 5]
```diff
@@ -136,3 +137,28 @@ def callback(event):
...
+                events.daemon.trigger('_test.event', None, callback)
+
+        with events.bound('_test.event',  '_test.handler', self._incrementWithResponse):
+            events.daemon.trigger('_test.event', {'amount': 2}, callback)
+
+        self.assertEqual(self.ctr, 3)
+        self.assertEqual(self.responses, ['foo'])
+
+        events.daemon.stop()
+        events.daemon = oldDaemon
```
Comment (mgrauer) [inline tests/cases/events_test.py]: Resetting state at the end makes me nervous.  Can you do the above testing work in a context manager and then reset the state outside the CM?

File: tests/cases/events_test.py [hunk 6]
```diff
@@ -136,3 +138,25 @@ def callback(event):
             self.assertEqual(events.daemon.eventQueue.qsize(), 0)
             self.assertEqual(self.ctr, 3)
             self.assertEqual(self.responses, ['foo'])
+            events.daemon.stop()
+
+    @mock.patch.object(events, 'daemon', new=events.ForegroundEventsDaemon())
+    def testForegroundDaemon(self):
```
Comment (mgrauer) [inline tests/cases/events_test.py]: This works for the context manager, thanks.

One last worry is that this doesn't actually test that we are using the ForegroundEventsDaemon for events.daemon. Am I right to worry about that?  Couldn't this test code also work if the standard async daemon is used?

I guess it comes down to trusting that the mock.patch works.  Or, trust, but verify.  Perhaps just a type check of events.daemon, that it is a ForegroundEventsDaemon?

**Comment by zachmullen** [comment]
@mgrauer any other issues or is this good to merge?

File: tests/cases/events_test.py (hunk 6 above)
Comment (zachmullen) [inline tests/cases/events_test.py]: > Am I right to worry about that? Couldn't this test code also work if the standard async daemon is used?

It would create a race condition, and this test would probably fail intermittently if the mock was failing. I added the assertion that the mock was working, there's certainly no harm in that.

**Comment by mgrauer** [comment]
LGTM.  I'm also getting a couple messages in the logs when I run the tests that tell me the right thing is occurring.  Hmmm, is that something we can test for?  

    WARNING: Handler "_test.handler" for event "_test.event" was triggered on the daemon, but is actually running synchronously.
